File: services/rag/src/embeddings/voyage.ts
import type { EmbeddingClient, EmbeddingConfig } from "./types.js";

export class VoyageEmbedding implements EmbeddingClient {
  private readonly apiKey: string;
  readonly model: string;
  readonly dimensions: number;
  private readonly baseUrl = "https://api.voyageai.com/v1";

  constructor(config: EmbeddingConfig) {
    this.apiKey = config.apiKey;
    this.model = config.model || "voyage-2";
    this.dimensions = config.dimensions ?? 1024;
  }

  async embed(texts: string[]): Promise<number[][]> {
    const response = await fetch(`${this.baseUrl}/embeddings`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: this.model,
        input: texts,
        input_type: "document",
      }),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Voyage embedding failed: ${error}`);
    }

    const data = (await response.json()) as {
      data: Array<{ embedding: number[] }>;
    };
    return data.data.map((item) => item.embedding);
  }

  async embedSingle(text: string): Promise<number[]> {
    const [embedding] = await this.embed([text]);
    return embedding;
  }
}

File: services/confluence-importer/src/confluence_importer/embeddings/voyage.py
"""Voyage AI embeddings via the REST API (no SDK dependency)."""

import httpx

from confluence_importer.embeddings.base import EmbeddingApiError, batched, validate_dimensions
from confluence_importer.retry import with_retry

_BASE_URL = "https://api.voyageai.com/v1"
_MAX_BATCH = 128


class VoyageEmbedding:
    def __init__(
        self,
        *,
        api_key: str,
        model: str,
        dimensions: int = 1024,
        http_client: httpx.Client | None = None,
    ) -> None:
        self.model = model or "voyage-2"
        self.dimensions = dimensions
        self._http = http_client or httpx.Client(
            headers={"Authorization": f"Bearer {api_key}"},
            timeout=60.0,
        )

    def embed_texts(self, texts: list[str]) -> list[list[float]]:
        results: list[list[float]] = []
        for batch in batched(texts, _MAX_BATCH):
            results.extend(self._embed_batch(batch))
        return results

    def embed_query(self, text: str) -> list[float]:
        return self.embed_texts([text])[0]

    def _embed_batch(self, texts: list[str]) -> list[list[float]]:
        def do_request() -> list[list[float]]:
            response = self._http.post(
                f"{_BASE_URL}/embeddings",
                json={"model": self.model, "input": texts, "input_type": "document"},
            )
            if response.status_code >= 400:
                raise EmbeddingApiError("Voyage", response.status_code, response.text)
            data = response.json()["data"]
            # The API is not guaranteed to preserve input order; restore it via
            # the per-item index when the response carries one.
            if all("index" in item for item in data):
                data = sorted(data, key=lambda item: item["index"])
            embeddings = [item["embedding"] for item in data]
            return validate_dimensions("Voyage", self.model, self.dimensions, embeddings)

        return with_retry(do_request)
